# 4.5 Data Consistency Checks

## Contents

### 1. Importing Libaries and Data

### 2. Creating a small test dataframe for practise


### 3. Missing Values

### 3.1 Finding Missing Values

### 3.2 Addressing Missing Values

### 3.3  Creating a new data frame with only non-missing values 

### 4. Duplicates

### 4.1 Finding Duplicates

### 4.2 Adressing Duplicates 

### 5. Task Data Consistency Check

# Script

### 1. Importing Libaries and Data

In [98]:
import pandas as pd
import numpy as np
import os

In [99]:
path = r'C:\Users\tinar\OneDrive\Dokumente\00 DATA ANALYST Carrer Foundry\04 Immersion Course\4. Python Fundamentals for Data Analysts\2209 Instacart Basket Analysis'

In [52]:
# products.csv data set from Original Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', '01 Original Data', '4.3_orders_products','products.csv'), index_col = False)

In [100]:
# orderswrangled.csv set from Prepared Data 
df_ords = pd.read_csv(os.path.join(path, '02 Data', '02 Prepared Data', 'orders_wrangled.csv'), index_col = False)

### 2. Creating a small test dataframe for practise

In [101]:
# create a dataframe
df_test = pd.DataFrame()

In [102]:
# create a miexed type column
df_test['mix'] = ['a', 'b', 1, True]

In [103]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [104]:
# function for checking whether a dataframe contains any mixed-type columns - The structure that’s being used in this code is called a “for-loop.

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [105]:
df_test['mix'] = df_test['mix'].astype('str')

In [106]:
df_test

mix
0     a
1     b
2     1
3  True

### 3. Missing Values 

### 3.1 Finding Missing Values

In [107]:
# sum function show how many total missing observations there are / 
# here the only column with missing values is the "product_name" column, and it’s missing 16 values.
df_prods.isnull().sum()

Unnamed: 0       0
product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [108]:
# To actually view these 16 values, you can create a subset of the dataframe containing only the values in question
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [109]:

df_nan

Empty DataFrame
Columns: [Unnamed: 0, product_id, product_name, aisle_id, department_id, prices]
Index: []

### 3.2 Addressing Missing Values

In [63]:
# Create a new variable that acts like a flag based on the missing value.
# Impute the value with the mean or median of the column (if the variable is numeric).
# Remove or filter out the missing data.

In [110]:
df_prods.median()

C:\Users\tinar\AppData\Local\Temp\ipykernel_12500\3387230063.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_prods.median()


Unnamed: 0       24851.5
product_id       24850.5
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [111]:
print(df_prods.mean(numeric_only=True)) # looked the error message up on the internet. Found the following solution:

Unnamed: 0       24852.005053
product_id       24850.349775
aisle_id            67.762442
department_id       11.728942
prices               9.993282
dtype: float64


In [112]:
# USE the following code to replace your missing values
# df['column with missings'].fillna(mean value, inplace=True) 
# df['column with missings'].fillna(median value, inplace=True)

In [113]:
df_prods.shape

(49672, 6)

### 3.3  Creating a new data frame with only non-missing values 

In [114]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [115]:
df_prods_clean.shape #new data frame now has not included the 16 missing values

(49672, 6)

###### Another way you can drop all missing values is via the following command:

In [116]:
df_prods.dropna(inplace = True)

In [117]:




#
#
df_prods.dropna(subset = ['product_name'], inplace = True)

In [118]:
# In both cases, rather than creating an entirely new dataframe, you’re overwriting 
# df_prods with a new version of df_prods that doesn’t contain the missing values. 
# This is done by way of the inplace = True function, which overwrites the original dataframe.

## 4. Duplicates

### 4.1 Finding Duplicates

In [119]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [120]:
df_dups #This will display all the duplicate rows within your dataframe.

Empty DataFrame
Columns: [Unnamed: 0, product_id, product_name, aisle_id, department_id, prices]
Index: []



### 4.2 Addressing Duplicates

In [121]:
df_prods_clean # First checking the current number of rows :49677 rows × 5 columns

Unnamed: 0  product_id  \
0               0           1   
1               1           2   
2               2           3   
3               3           4   
4               4           5   
...           ...         ...   
49667       49688       49684   
49668       49689       49685   
49669       49690       49686   
49670       49691       49687   
49671       49692       49688   

                                            product_name  aisle_id  \
0                             Chocolate Sandwich Cookies        61   
1                                       All-Seasons Salt       104   
2                   Robust Golden Unsweetened Oolong Tea        94   
3      Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4                              Green Chile Anytime Sauce         5   
...                                                  ...       ...   
49667          Vodka, Triple Distilled, Twist of Vanilla       124   
49668                 En Croute Roast Hazelnut Cranberry        42   
49669                                   Artisan Baguette       112   
49670         Smartblend Healthy Metabolism Dry Cat Food        41   
49671                             Fresh Foaming Cleanser        73   

       department_id  prices  
0                 19     5.8  
1                 13     9.3  
2                  7     4.5  
3                  1    10.5  
4                 13     4.3  
...              ...     ...  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5  

[49672 rows x 6 columns]

In [122]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates() # This command has created a new dataframe—df_prods_clean_no_dups that contains only the unique rows from df_prods_clean.

In [123]:

df_prods_clean_no_dups #now containing 5 rows less (the duplicates)

Unnamed: 0  product_id  \
0               0           1   
1               1           2   
2               2           3   
3               3           4   
4               4           5   
...           ...         ...   
49667       49688       49684   
49668       49689       49685   
49669       49690       49686   
49670       49691       49687   
49671       49692       49688   

                                            product_name  aisle_id  \
0                             Chocolate Sandwich Cookies        61   
1                                       All-Seasons Salt       104   
2                   Robust Golden Unsweetened Oolong Tea        94   
3      Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4                              Green Chile Anytime Sauce         5   
...                                                  ...       ...   
49667          Vodka, Triple Distilled, Twist of Vanilla       124   
49668                 En Croute Roast Hazelnut Cranberry        42   
49669                                   Artisan Baguette       112   
49670         Smartblend Healthy Metabolism Dry Cat Food        41   
49671                             Fresh Foaming Cleanser        73   

       department_id  prices  
0                 19     5.8  
1                 13     9.3  
2                  7     4.5  
3                  1    10.5  
4                 13     4.3  
...              ...     ...  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5  

[49672 rows x 6 columns]

### Exporting Dataframe

In [124]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','02 Prepared Data','products_checked.csv'))

# 5. Task 4.5  Data Consistency Checks

In [125]:
df_prods = pd.read_csv(os.path.join(path, '02 Data','02 Prepared Data','products_checked.csv'), index_col = False)

### Step 1 consistency checks on df_prods - done! 

### Step 2  interpret the output of this function

In [126]:
df_prods.describe()

Unnamed: 0.1    Unnamed: 0    product_id      aisle_id  department_id  \
count  49672.000000  49672.000000  49672.000000  49672.000000   49672.000000   
mean   24835.500000  24852.005053  24850.349775     67.762442      11.728942   
std    14339.215622  14342.265579  14340.705287     38.315784       5.850779   
min        0.000000      0.000000      1.000000      1.000000       1.000000   
25%    12417.750000  12432.750000  12432.750000     35.000000       7.000000   
50%    24835.500000  24851.500000  24850.500000     69.000000      13.000000   
75%    37253.250000  37272.250000  37268.250000    100.000000      17.000000   
max    49671.000000  49692.000000  49688.000000    134.000000      21.000000   

             prices  
count  49672.000000  
mean       9.993282  
std      453.615536  
min        1.000000  
25%        4.100000  
50%        7.100000  
75%       11.100000  
max    99999.000000

The max price of $99999 seems pretty high! 

### Step 3 Check for mixed-type data in your df_ords dataframe.

In [127]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

The output is not showing any "mix". 

### Step 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.



Didn't find any but in case I would have the code is the following: df_ords['mix'] = df_ords['mix'].astype('str')

### Step 5 Run a check for missing values in your df_ords dataframe

In [128]:
# sum function show how many total missing observations there are 
# here the only column with missing values is the "days_since_prior_order" column, and it’s missing 206209 values.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Here the only column with missing values is the "days_since_prior_order" column, and it’s missing 206209 values.

In [129]:
df_ords.shape

(3421083, 8)

### Step 6: Address the missing values using an appropriate method.

In [130]:

# We have 206209 missing values out of 3421083 = ca. 6% 

In [131]:
print(df_ords.mean(numeric_only=True))

Unnamed: 0                1.710541e+06
order_id                  1.710542e+06
user_id                   1.029782e+05
order_number              1.715486e+01
order_dow                 2.776219e+00
order_hour_of_day         1.345202e+01
days_since_prior_order    1.111484e+01
dtype: float64


In [132]:
print(df_ords.median(numeric_only=True))

Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
order_dow                       3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64



After checking the mean (average) and the median (middle number) I would address the missing data by replacing them with the average (1.111). 

In [87]:
# df['column with missings'].fillna(mean value, inplace=True) 
df_ords['days_since_prior_order'].fillna(1.111484e+01, inplace=True)

###  Step 7: Run a check for duplicate values in your df_ords data

In [133]:
df_dups = df_ords[df_ords.duplicated()]

In [134]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []


### Step 8: Address the duplicates using an appropriate method. 

No duplicates 

In [135]:
df_prods.to_csv(os.path.join(path, '02 Data','02 Prepared Data','1_consistent_product.csv'), index=False)

In [136]:
df_ords.to_csv(os.path.join(path, '02 Data','02 Prepared Data','1_consistent_orders.csv'), index = False)

## Checking Data for 4.10 

In [ ]:
# customers.csv data set from Original Data
df_custo = pd.read_csv(os.path.join(path, '02 Data', '01 Original Data', 'customers','customers.csv'), index_col = False)

In [ ]:
#customer original data

In [ ]:
df_custo

In [91]:
df_dups = df_custo[df_custo.duplicated()]

In [92]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Surnam, Gender, STATE, Age, date_joined, n_dependants, fam_status, income]
Index: []

In [93]:
df_custo.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64